sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollectCTEDefinitions.scala

```diff
@@ -282,11 +283,15 @@ object CTESubstitution extends Rule[LogicalPlan] {
           if (alwaysInline) {
             d.child
           } else {
-            // Add a `SubqueryAlias` for hint-resolving rules to match relation names.
-            SubqueryAlias(table, CTERelationRef(d.id, d.resolved, d.output, d.isStreaming))
+            // Add unresolved with CTE relations to the plan and we
+            // will do CTE resolution later in analyzer based on this node.
+            UnresolvedWithCTERelations(u, cteRelations)
           }
         }.getOrElse(u)
 
+      case p: PlanWithUnresolvedIdentifier =>
```

Comment:
Let's keep the rule name unchanged and only make this change here. We should also add comments to explain it
```
// We must look up CTE relations first when resolving `UnresolvedRelation`s, but we can't do it here
// as `PlanWithUnresolvedIdentifier` is a leaf node and may produce `UnresolvedRelation` later. Here
// we wrap it with `UnresolvedWithCTERelations` so that we can delay the CTE relations lookup after
// `PlanWithUnresolvedIdentifier` is resolved.
```

Reply:
Another thing is that we should guarantee the `UnresolvedRelation` should be resolved by CTE relations lookup before the normal table lookup path. My proposal: `UnresolvedWithCTERelations` should be a leaf node so that the normal table lookup rule can't transform the `UnresolvedRelation` inside it. `ResolveIdentifierClause` should handle `UnresolvedWithCTERelations` specially and resolve `PlanWithUnresolvedIdentifier` inside it.

Reply:
I have made changes suggested in these comments, hopefully I understood correctly what you suggested, please review :)